### PR TITLE
feat(auth): Expose AuthConfiguration directly

### PR DESCRIPTION
**Pull request by mattcreaser**
- [x] PR title and description conform to [Pull Request](https://github.com/aws-amplify/amplify-android/blob/main/CONTRIBUTING.md#pull-request-guidelines) guidelines.

*Issue #, if available:*

*Description of changes:*
Amplify previously exposed raw JSON via `getPluginConfiguration` for Authenticator to consume. This change instead exposed a pre-parsed `AuthConfiguration`.

- Added `signupAttributes`, `usernameAttributes`, `verificationMechanisms`, and `passwordProtectionSettings` to `AuthConfiguration`
- Exposed `AuthConfiguration` and related types with `@InternalAmplifyApi` annotations.
- Deprecated existing `getPluginConfiguration` function.
- Moved `AuthConfiguration` to `com.amplifyframework.auth.cognito` package. This is a more appropriate location for this class.

Equivalent Swift PR is here: https://github.com/aws-amplify/amplify-swift/pull/3566

*How did you test these changes?*
- Tested consuming new internal API in Authenticator 

*Documentation update required?*
- [x] No
- [ ] Yes (Please include a PR link for the documentation update)

*General Checklist*
- [ ] Added Unit Tests
- [ ] Added Integration Tests
- [ ] Security oriented best practices and standards are followed (e.g. using input sanitization, principle of least privilege, etc)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt [hunk 1]
```diff
@@ -64,5 +110,24 @@ internal data class AuthConfiguration internal constructor(
...
         }
+
+        private fun getPasswordProtectionSettings(authConfig: JSONObject?): PasswordProtectionSettings? {
+            val passwordSettings = authConfig?.optJSONObject("passwordProtectionSettings") ?: return null
+            val passwordLength = passwordSettings.optInt("passwordPolicyMinLength")
+            val passwordRequirements = passwordSettings.optJSONArray("passwordPolicyCharacters")?.map {
+                getString(it)
+            } ?: emptyList()
+            return PasswordProtectionSettings(
+                length = passwordLength,
```
Comment (lawmicha) [inline aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt:121]: what happens on `passwordSettings.optInt("passwordPolicyMinLength")`  when the JSONObject does not contain the key-value? Does it return null for `passwordLength`, if so what happens when creating the PasswordProtectionSettings's `length` with null?

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt [hunk 2]
```diff
@@ -15,28 +15,70 @@
...
+}
+
+@InternalAmplifyApi
+data class PasswordProtectionSettings(
+    @IntRange(from = 6, to = 99) val length: Int,
+    val requiresNumber: Boolean,
+    val requiresSpecial: Boolean,
+    val requiresUpper: Boolean,
+    val requiresLower: Boolean
+)
```
Comment (lawmicha) [inline aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt:39]: nice, i like that this data type is closer to the new config schema's type. I had opted to make less changes in the Authenticator by just moving over the Authenticator's PasswordProtectionSettings type over to Amplify, so the difference is this https://github.com/aws-amplify/amplify-swift/pull/3566/files#diff-8cfb4aa001ce3e3a0d142cf84c5371705975660e54b372128d159115ea425220

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt (hunk 1 above)
Comment (mattcreaser) [reply aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt:113]: `optInt` returns 0 if the field is missing or is not a number

File: aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt (hunk 2 above)
Comment (mattcreaser) [reply aws-auth-cognito/src/main/java/com/amplifyframework/statemachine/codegen/data/AuthConfiguration.kt:34]: I also moved this directly (just renamed the class) - on Android we were already using a class of this structure 🤷

**Comment by tylerjroach** [comment]
Looks good, but thoughts on removing the Java style builders since those are all internal?

**Comment by mattcreaser** [comment]
The builders are only used from tests and could be removed, but I'd do that separately to avoid bloating this PR.

**Review by tylerjroach — APPROVED** [review]

**Review by lawmicha — APPROVED** [review]

**Pull request merged**